File: Infart/Background/Grattacielo.cs
using Infart.Drawing;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace Infart.Background
{
    public class Grattacielo : GameObject
    {
        private readonly int _h;

        private readonly int _oneBlockHeight = 25;

        private readonly int _wGrattacielo;

        private Rectangle _collisionRectangle;

        private readonly Rectangle _textureRectangle;

        private readonly Texture2D _textureReference;

        public Grattacielo(Rectangle textureRectangle, Texture2D textureReference)
        {
            Position = Vector2.Zero;

            _wGrattacielo = textureRectangle.Width;
            _h = textureRectangle.Height / _oneBlockHeight;
            Origin = new Vector2(0, textureRectangle.Height);
            _collisionRectangle.Width = textureRectangle.Width;
            _collisionRectangle.Height = textureRectangle.Height;

            _textureReference = textureReference;
            _textureRectangle = textureRectangle;
        }

        public void Move(Vector2 amount)
        {
            base.Position += amount;
        }

        public override Rectangle CollisionRectangle
        {
            get
            {
                return _collisionRectangle;
            }
        }

        public override Vector2 Position
        {
            get
            {
                return base.Position;
            }
            set
            {
                _collisionRectangle.X = (int)value.X;
                _collisionRectangle.Y = (int)(value.Y - Origin.Y);
                base.Position = value;
            }
        }

        public int Width
        {
            get { return _textureRectangle.Width; }
        }

        public int Height
        {
            get { return _textureRectangle.Height; }
        }

        public int HeightInBlocksNumber
        {
            get { return _h; }
        }

        public Vector2 PositionAtTopLeftCorner()
        {
            return new Vector2(
                Position.X,
                Position.Y - Origin.Y);
        }

        public override void Update(double gameTime)
        {
        }

        public override void Draw(SpriteBatch spriteBatch)
        {
            spriteBatch.Draw(
                _textureReference,
                base.Position,
                _textureRectangle,
                Color.White,
                0.0f,
                Origin,
                Scale,
                Flip,
                Depth);
        }
    }
}

File: Web/src/background/Grattacielo.ts
import {Rectangle, Sprite, Texture} from "pixi.js";
import Camera from "../world/Camera.ts";

class Grattacielo {

    private readonly _sprite: Sprite;

    constructor(
        texture: Texture,
        camera: Camera) {

        this._sprite = new Sprite(texture);
        camera.addToWorld(this._sprite);
        this._sprite.anchor.set(0, 1);
        this._sprite.y = 0;
    }

    get width() {
        return this._sprite.width;
    }

    get height() {
        return this._sprite.height;
    }

    get x() {
        return this._sprite.x;
    }

    set x(x: number) {
        this._sprite.x = x;
    }

    get y() {
        return this._sprite.y;
    }

    set y(y: number) {
        this._sprite.y = y;
    }

    get collisionRectangle() {
        return new Rectangle(
            this._sprite.x,
            this._sprite.y - this._sprite.height,
            this._sprite.width,
            this._sprite.height);
    }

}

export default Grattacielo;